candle-kernels/src/cast.cu

```diff
@@ -24,19 +24,42 @@ extern "C" __global__ void FN_NAME( \
     } \
 } \
 
+#define CAST_BF_OP(SRC_TYPENAME, DST_TYPENAME, FN_NAME) \
+extern "C" __global__ void FN_NAME( \
+    const size_t numel, \
+    const size_t num_dims, \
+    const size_t *info, \
+    const SRC_TYPENAME *inp, \
+    DST_TYPENAME *out \
+) { \
+    const size_t *dims = info; \
+    const size_t *strides = info + num_dims; \
+    if (is_contiguous(num_dims, dims, strides)) { \
+        for (unsigned int i = blockIdx.x * blockDim.x + threadIdx.x; i < numel; i += blockDim.x * gridDim.x) { \
+            out[i] = (DST_TYPENAME) (float) inp[i]; \
```

Comment:
Can't find anything better, happy to try things.

The low level cast just doesn't exist. This seems OK since f32 >> (bf16, f16) so the uplift shouldn't be an issue.
I'm wondering how much the compiler can remove the intermediary conversion though.

Reply:
Ok let's go ahead with this version then.